### PR TITLE
FEATURE: SD-446 Unify behavior of stackMeasuresToPercent and computeRatio on Combo chart

### PR DESCRIPTION
**Pull request by nadtan**
---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist
- [x] Squash commits
- [x] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [x] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [x] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [x] Successful `extended test - examples`
- [x] Successful `extended test - storybook`
- [x] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)

# Related PRs
- gdc-analytical-designer: https://github.com/gooddata/gdc-analytical-designer/pull/2236
- gdc-dashboards: https://github.com/gooddata/gdc-dashboards/pull/1965

# Related Jira tasks

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3524/ : SUCCESS in 4m 03s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4369/ : SUCCESS in 15s

**Comment by nadtan** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1877/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1220/ : SUCCESS in 12m 16s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4370/ : SUCCESS in 6s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3526/ : SUCCESS in 8m 00s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4373/ : SUCCESS in 7s

**Comment by nadtan** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-970.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1228/ : SUCCESS in 37m 50s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4388/ : SUCCESS in 8s

**Comment by ptquang86** [comment]
LGTM

File: src/components/tests/ComboChart.spec.tsx [hunk 1]
```diff
@@ -64,4 +66,76 @@ describe("ComboChart", () => {
...
+            const wrapper = renderChart([M5WithRatio], [], config);
+            const configProps = wrapper.find(AfmComboChart).prop("config");
+
+            expect(configProps.stackMeasures).toBeFalsy();
+            expect(configProps.stackMeasuresToPercent).toBeFalsy();
+        });
+
+        it.each([["primary", [M5, M5WithRatio], []], ["secondary", [], [M5, M5WithRatio]]])(
+            "should ignore computeRatio when %s measure bucket has multiple items",
+            (_name: string, primaryMeasures, secondaryMeasures) => {
```
Comment (naibinh) [inline src/components/tests/ComboChart.spec.tsx]: we should add type to two last params

File: src/helpers/optionalStacking/test/common.spec.ts [hunk 2]
```diff
@@ -2,61 +2,72 @@
...
-            {
-                localIdentifier: STACK,
-                items: props.stackBy ? [props.stackBy] : [],
-            },
-        ];
-    }
+const A1: VisualizationObject.IVisualizationAttribute = attribute("a1").localIdentifier("a1");
+const A2: VisualizationObject.IVisualizationAttribute = attribute("a2").localIdentifier("a2");
+const M1: VisualizationObject.IMeasure = measure("m1").localIdentifier("m1");
+const M2: VisualizationObject.IMeasure = measure("m2").localIdentifier("m2");
```
Comment (naibinh) [inline src/helpers/optionalStacking/test/common.spec.ts]: how about this?
```
const [A1, A2]: Array<VisualizationObject.IVisualizationAttribute> = ["a1", "a2"].map(...);
// same for M1, M2
```

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3571/ : SUCCESS in 23m 21s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4431/ : SUCCESS in 11s

File: src/components/tests/ComboChart.spec.tsx (hunk 1 above)
Comment (nadtan) [reply src/components/tests/ComboChart.spec.tsx]: updated

**Comment by naibinh** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3590/ : SUCCESS in 12m 30s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4469/ : SUCCESS in 13s

File: src/helpers/dimensions.ts [hunk 3]
```diff
@@ -11,6 +11,13 @@ export function findBucketByLocalIdentifier(buckets: VisualizationObject.IBucket
     return (buckets || []).find(bucket => bucket.localIdentifier === bucketName);
 }
 
+export function getBucketItems(
```
Comment (ivan-nejezchleb) [inline src/helpers/dimensions.ts]: I would like to remove this helper and 'findBucketByLocalIdentifier' to some other file and keep here only functions around dimensions. After merge of app-components to react-components, there is file bucketHelper.ts, BUT it works with different interfaces (with UI buckets).
So probably something like 'mdObjBucketHelper' would be good place for these two helpers

File: src/helpers/optionalStacking/common.ts [hunk 4]
```diff
@@ -23,70 +20,44 @@ export function getViewByTwoAttributes(
     return [viewBy] as VisualizationObject.IVisualizationAttribute[];
 }
 
-export interface ISanitizedBucketsAndStackingConfig {
-    buckets: VisualizationObject.IBucket[];
-    config: IChartConfig;
+/**
+ * multiple measures => ignore computeRatio
+ */
+export function sanitizeMeasures(
```
Comment (ivan-nejezchleb) [inline src/helpers/optionalStacking/common.ts]: sanitizeMeasures seems to me too generic name. If it sanitizes only 'ComputeRatio' I would call it 'SanitizeComputeRatioOnMeasures'

File: src/helpers/optionalStacking/test/common.spec.ts [hunk 5]
```diff
@@ -2,138 +2,71 @@
...
-            viewBy: [A1],
-        });
-        const config: IChartConfig = {
-            stackMeasures: true,
-            stackMeasuresToPercent: true,
-        };
-        const sanitizedBucketsAndStackingConfig: ISanitizedBucketsAndStackingConfig = getSanitizedBucketsAndStackingConfig(
-            buckets,
-            config,
+    it("should sanitized computeRatio if there are multiple measures", () => {
```
Comment (ivan-nejezchleb) [inline src/helpers/optionalStacking/test/common.spec.ts]: missing test case for one measure and second "force" param = true

File: src/components/ComboChart.tsx [hunk 6]
```diff
@@ -52,15 +54,24 @@ export function ComboChart(props: IComboChartProps): JSX.Element {
     }
 
     const { primaryMeasures = [], secondaryMeasures = [] } = clonedProps;
```
Comment (ivan-nejezchleb) [inline src/components/ComboChart.tsx]: whole this preparation of measures starts to be quite big, I would like to separate it into extra function(s)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3676/ : SUCCESS in 11m 25s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4587/ : SUCCESS in 11s

File: src/components/ComboChart.tsx [hunk 7]
```diff
@@ -14,13 +14,14 @@ import { convertBucketsToAFM, convertBucketsToMdObject } from "../helpers/conver
...
-import { sanitizeConfig, sanitizeMeasures } from "../helpers/optionalStacking/common";
+import { sanitizeConfig, sanitizeComputeRatioOnMeasures } from "../helpers/optionalStacking/common";
+import { IChartConfig } from "..";
 
 export interface IComboChartBucketProps {
-    columnMeasures?: VisualizationInput.IMeasure[];
-    lineMeasures?: VisualizationInput.IMeasure[];
-    primaryMeasures?: VisualizationInput.IMeasure[];
-    secondaryMeasures?: VisualizationInput.IMeasure[];
+    columnMeasures?: VisualizationInput.AttributeOrMeasure[];
```
Comment (ivan-nejezchleb) [inline src/components/ComboChart.tsx]: why this change?

Comment (nadtan) [reply src/components/ComboChart.tsx]: To make it consistency with other visualizations (eg Column/Bar....)
And to get rid of typescript error because `sanitizeComputeRatioOnMeasures` takes `BucketItem` as parameter and it covers for Column/Bar/Area charts as well.

Comment (nadtan) [reply src/components/ComboChart.tsx]: Reverted

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3756/ : SUCCESS in 4m 21s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4723/ : SUCCESS in 6s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3762/ : SUCCESS in 13m 16s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4735/ : SUCCESS in 8s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2042/ : SUCCESS in 10m 32s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3769/ : SUCCESS in 15m 26s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1002/ : SUCCESS in 24s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1124/ : SUCCESS in 30m 30s
